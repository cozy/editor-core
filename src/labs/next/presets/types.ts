import { FeatureFlags } from '../../../plugins/feature-flags-context/types';

export interface EditorPresetProps {
  excludes?: Set<string>;
  experimental?: Array<string>;
  featureFlags?: FeatureFlags;
}
