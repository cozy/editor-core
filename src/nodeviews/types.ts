import type { Node as PMNode } from '@atlaskit/editor-prosemirror/model';

export interface ReactNodeProps {
  selected: boolean;
}

export type ProsemirrorGetPosHandler = () => number | undefined;
export type getPosHandler = getPosHandlerNode | boolean;
export type getPosHandlerNode = () => number | undefined;
export type ReactComponentProps = { [key: string]: unknown };
export type ForwardRef = (node: HTMLElement | null) => void;
export type shouldUpdate = (nextNode: PMNode) => boolean;
