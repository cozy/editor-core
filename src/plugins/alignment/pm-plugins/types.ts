export type AlignmentState = 'start' | 'end' | 'center';
export type AlignmentPluginState = {
  align: AlignmentState;
  isEnabled?: boolean;
};
