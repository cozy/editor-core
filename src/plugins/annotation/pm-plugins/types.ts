import type { DecorationSet } from '@atlaskit/editor-prosemirror/view';
import type {
  EditorState,
  SelectionBookmark,
} from '@atlaskit/editor-prosemirror/state';
import type { PortalProviderAPI } from '@atlaskit/editor-common/portal-provider';
import type { EditorAnalyticsAPI } from '@atlaskit/editor-common/analytics';
import type {
  Dispatch,
  EventDispatcher,
} from '@atlaskit/editor-common/event-dispatcher';
import type { InlineCommentAnnotationProvider, AnnotationInfo } from '../types';

export enum ACTIONS {
  UPDATE_INLINE_COMMENT_STATE,
  SET_INLINE_COMMENT_DRAFT_STATE,
  INLINE_COMMENT_UPDATE_MOUSE_STATE,
  INLINE_COMMENT_CLEAR_DIRTY_MARK,
  ADD_INLINE_COMMENT,
  INLINE_COMMENT_SET_VISIBLE,
  CLOSE_COMPONENT,
  SET_SELECTED_ANNOTATION,
}

export interface InlineCommentPluginOptions {
  dispatch: Dispatch;
  eventDispatcher: EventDispatcher;
  portalProviderAPI: PortalProviderAPI;
  provider: InlineCommentAnnotationProvider;
  editorAnalyticsAPI: EditorAnalyticsAPI | undefined;
}
export interface InlineCommentMouseData {
  isSelecting: boolean;
}

export type InlineCommentMap = { [key: string]: boolean };
export type InlineCommentAction =
  | {
      type: ACTIONS.UPDATE_INLINE_COMMENT_STATE;
      data: InlineCommentMap;
    }
  | {
      type: ACTIONS.SET_INLINE_COMMENT_DRAFT_STATE;
      data: {
        drafting: boolean;
        editorState: EditorState;
      };
    }
  | {
      type: ACTIONS.INLINE_COMMENT_UPDATE_MOUSE_STATE;
      data: {
        mouseData: InlineCommentMouseData;
      };
    }
  | { type: ACTIONS.INLINE_COMMENT_CLEAR_DIRTY_MARK }
  | { type: ACTIONS.CLOSE_COMPONENT }
  | {
      type: ACTIONS.ADD_INLINE_COMMENT;
      data: {
        drafting: boolean;
        inlineComments: InlineCommentMap;
        editorState: EditorState;
        selectedAnnotations: AnnotationInfo[];
      };
    }
  | { type: ACTIONS.INLINE_COMMENT_SET_VISIBLE; data: { isVisible: boolean } }
  | {
      type: ACTIONS.SET_SELECTED_ANNOTATION;
      data: { selectedAnnotations: AnnotationInfo[] };
    };

export type InlineCommentPluginState = {
  annotations: InlineCommentMap;
  selectedAnnotations: AnnotationInfo[];
  dirtyAnnotations?: boolean;
  mouseData: InlineCommentMouseData;
  draftDecorationSet?: DecorationSet;
  bookmark?: SelectionBookmark;

  // Denotes if annotations are allowed to be create on empty nodes or nodes of whitespace (Confluence spec)
  disallowOnWhitespace: boolean;

  // Allow users to hide inline comments during editing
  isVisible: boolean;
  skipSelectionHandling: boolean;
};
