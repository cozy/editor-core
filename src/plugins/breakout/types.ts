import type { ContentNodeWithPos } from '@atlaskit/editor-prosemirror/utils';
export interface BreakoutPluginState {
  breakoutNode: ContentNodeWithPos | undefined;
}
