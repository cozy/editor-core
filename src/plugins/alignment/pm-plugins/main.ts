import { SafePlugin } from '@atlaskit/editor-common/safe-plugin';
import type { EditorState } from '@atlaskit/editor-prosemirror/state';
import { PluginKey } from '@atlaskit/editor-prosemirror/state';
import type { Dispatch } from '../../../event-dispatcher';
import { isAlignable } from '../commands';
import { getActiveAlignment } from '../utils';
import type { AlignmentPluginState } from './types';

export function createInitialPluginState(
  editorState: EditorState,
  pluginConfig: AlignmentPluginState,
): AlignmentPluginState {
  return {
    align: getActiveAlignment(editorState) || pluginConfig.align,
    isEnabled: true,
  };
}

export const pluginKey = new PluginKey<AlignmentPluginState>('alignmentPlugin');

export function createPlugin(
  dispatch: Dispatch,
  pluginConfig: AlignmentPluginState,
): SafePlugin {
  return new SafePlugin({
    key: pluginKey,
    state: {
      init(_config, editorState) {
        return createInitialPluginState(editorState, pluginConfig);
      },
      apply(_tr, state: AlignmentPluginState, _prevState, nextState) {
        const nextPluginState = getActiveAlignment(nextState)!;
        const isEnabled = isAlignable(nextPluginState)(nextState);
        const newState = {
          ...state,
          align: nextPluginState,
          isEnabled,
        };
        if (nextPluginState !== state.align || isEnabled !== state.isEnabled) {
          dispatch(pluginKey, newState);
        }
        return newState;
      },
    },
  });
}
