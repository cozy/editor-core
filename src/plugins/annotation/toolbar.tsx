import React from 'react';
import type { IntlShape } from 'react-intl-next';
import type { EditorState } from '@atlaskit/editor-prosemirror/state';
import type { NodeType } from '@atlaskit/editor-prosemirror/model';
import CommentIcon from '@atlaskit/icon/glyph/comment';
import type { Command } from '../../types';
import { addInlineComment, ToolTipContent } from '../../keymaps';
import type {
  FloatingToolbarConfig,
  FloatingToolbarButton,
} from '@atlaskit/editor-common/types';
import type { EditorAnalyticsAPI } from '@atlaskit/editor-common/analytics';
import { setInlineCommentDraftState } from './commands';
import { AnnotationTestIds, AnnotationSelectionType } from './types';
import { isSelectionValid } from './utils';
import { annotationMessages } from '@atlaskit/editor-common/messages';
import {
  calculateToolbarPositionAboveSelection,
  calculateToolbarPositionTrackHead,
} from '@atlaskit/editor-common/utils';

export const buildToolbar =
  (editorAnalyticsAPI: EditorAnalyticsAPI | undefined) =>
  (
    state: EditorState,
    intl: IntlShape,
    isToolbarAbove: boolean = false,
  ): FloatingToolbarConfig | undefined => {
    const { schema } = state;
    const selectionValid = isSelectionValid(state);
    if (selectionValid === AnnotationSelectionType.INVALID) {
      return undefined;
    }

    const createCommentMessage = intl.formatMessage(
      annotationMessages.createComment,
    );
    const commentDisabledMessage = intl.formatMessage(
      annotationMessages.createCommentInvalid,
    );

    const createComment: FloatingToolbarButton<Command> = {
      type: 'button',
      showTitle: true,
      disabled: selectionValid === AnnotationSelectionType.DISABLED,
      testId: AnnotationTestIds.floatingToolbarCreateButton,
      icon: CommentIcon,
      tooltipContent:
        selectionValid === AnnotationSelectionType.DISABLED ? (
          commentDisabledMessage
        ) : (
          <ToolTipContent
            description={createCommentMessage}
            keymap={addInlineComment}
          />
        ),
      title: createCommentMessage,
      onClick: (state, dispatch) => {
        return setInlineCommentDraftState(editorAnalyticsAPI)(true)(
          state,
          dispatch,
        );
      },
    };

    const { annotation } = schema.marks;
    const validNodes = Object.keys(schema.nodes).reduce<NodeType[]>(
      (acc, current) => {
        const type = schema.nodes[current];
        if (type.allowsMarkType(annotation)) {
          acc.push(type);
        }
        return acc;
      },
      [],
    );

    const toolbarTitle = intl.formatMessage(annotationMessages.toolbar);
    const calcToolbarPosition = isToolbarAbove
      ? calculateToolbarPositionAboveSelection
      : calculateToolbarPositionTrackHead;
    const onPositionCalculated = calcToolbarPosition(toolbarTitle);

    return {
      title: toolbarTitle,
      nodeType: validNodes,
      items: [createComment],
      onPositionCalculated,
    };
  };
