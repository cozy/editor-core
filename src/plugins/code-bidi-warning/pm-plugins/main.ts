import { SafePlugin } from '@atlaskit/editor-common/safe-plugin';
import { codeBidiWarningMessages } from '@atlaskit/editor-common/messages';

import type { EditorProps, PMPluginFactoryParams } from '../../../types';

import { codeBidiWarningPluginKey } from '../plugin-key';

import {
  createBidiWarningsDecorationSetFromDoc,
  createPluginState,
  getPluginState,
} from './plugin-factory';

export const createPlugin = (
  { dispatch, getIntl }: PMPluginFactoryParams,
  { appearance }: { appearance: EditorProps['appearance'] },
) => {
  const intl = getIntl();

  const codeBidiWarningLabel = intl.formatMessage(
    codeBidiWarningMessages.label,
  );

  return new SafePlugin({
    key: codeBidiWarningPluginKey,
    state: createPluginState(dispatch, (state) => {
      // The appearance being mobile indicates we are in an editor being
      // rendered by mobile bridge in a web view.
      // The tooltip is likely to have unexpected behaviour there, with being cut
      // off, so we disable it. This is also to keep the behaviour consistent with
      // the rendering in the mobile Native Renderer.
      const tooltipEnabled = appearance !== 'mobile';

      return {
        decorationSet: createBidiWarningsDecorationSetFromDoc({
          doc: state.doc,
          codeBidiWarningLabel,
          tooltipEnabled,
        }),
        codeBidiWarningLabel,
        tooltipEnabled,
      };
    }),
    props: {
      decorations: (state) => {
        const { decorationSet } = getPluginState(state);

        return decorationSet;
      },
    },
  });
};
