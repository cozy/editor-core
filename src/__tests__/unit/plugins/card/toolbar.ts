import { ProviderFactory } from '@atlaskit/editor-common';
import { IntlProvider } from 'react-intl';
import createEditorFactory from '@atlaskit/editor-test-helpers/create-editor';
import {
  doc,
  p,
  inlineCard,
  blockCard,
  embedCard,
} from '@atlaskit/editor-test-helpers/schema-builder';
import { floatingToolbar } from '../../../../plugins/card/toolbar';
import { pluginKey } from '../../../../plugins/card/pm-plugins/main';

import commonMessages, {
  linkToolbarMessages,
  linkMessages,
} from '../../../../messages';

import RemoveIcon from '@atlaskit/icon/glyph/editor/remove';
import OpenIcon from '@atlaskit/icon/glyph/shortcut';
import UnlinkIcon from '@atlaskit/icon/glyph/editor/unlink';

import { FloatingToolbarButton } from '../../../../plugins/floating-toolbar/types';
import { Command } from '../../../../types';
import { getToolbarItems } from '../floating-toolbar/_helpers';

describe('card', () => {
  const createEditor = createEditorFactory();
  const providerFactory = new ProviderFactory();

  const editor = (doc: any) => {
    return createEditor({
      doc,
      providerFactory,
      editorProps: {
        UNSAFE_cards: {
          allowBlockCards: true,
          allowEmbeds: true,
          allowResizing: true,
        },
      },
      pluginKey,
    });
  };

  describe('toolbar', () => {
    const intlProvider = new IntlProvider({ locale: 'en' });
    const { intl } = intlProvider.getChildContext();

    const visitTitle = intl.formatMessage(linkMessages.openLink);
    const unlinkTitle = intl.formatMessage(linkToolbarMessages.unlink);
    const removeTitle = intl.formatMessage(commonMessages.remove);

    it('displays toolbar items in correct order for inlineCard', () => {
      const { editorView } = editor(
        doc(
          p(
            '{<node>}',
            inlineCard({
              url: 'http://www.atlassian.com/',
            })(),
          ),
        ),
      );

      const toolbar = floatingToolbar({
        allowBlockCards: true,
        allowEmbeds: true,
        allowResizing: true,
      })(editorView.state, intl, providerFactory);
      const toolbarItems = getToolbarItems(toolbar!, editorView);
      expect(toolbar).toBeDefined();
      expect(toolbarItems).toHaveLength(7);
      expect(toolbarItems).toMatchSnapshot();
    });

    it('displays toolbar items in correct order for blockCard', () => {
      const { editorView } = editor(
        doc(
          '{<node>}',
          blockCard({
            url: 'http://www.atlassian.com/',
          })(),
        ),
      );

      const toolbar = floatingToolbar({
        allowBlockCards: true,
        allowEmbeds: true,
        allowResizing: true,
      })(editorView.state, intl, providerFactory);
      const toolbarItems = getToolbarItems(toolbar!, editorView);
      expect(toolbar).toBeDefined();
      expect(toolbarItems).toHaveLength(7);
      expect(toolbarItems).toMatchSnapshot();
    });

    it('displays toolbar items in correct order for embedCard', () => {
      const { editorView } = editor(
        doc(
          '{<node>}',
          embedCard({
            url: 'http://www.atlassian.com/',
            layout: 'center',
          })(),
        ),
      );

      const toolbar = floatingToolbar({
        allowBlockCards: true,
        allowEmbeds: true,
        allowResizing: true,
      })(editorView.state, intl, providerFactory);
      const toolbarItems = getToolbarItems(toolbar!, editorView);
      expect(toolbar).toBeDefined();
      expect(toolbarItems).toHaveLength(14);
      expect(toolbarItems).toMatchSnapshot();
    });

    it('has an unlink button for inlineCard', () => {
      const { editorView } = editor(
        doc(
          p(
            '{<node>}',
            inlineCard({
              url: 'http://www.atlassian.com/',
            })(),
          ),
        ),
      );

      const toolbar = floatingToolbar({})(
        editorView.state,
        intl,
        providerFactory,
      );
      expect(toolbar).toBeDefined();

      const unlinkButton = getToolbarItems(toolbar!, editorView).find(
        item => item.type === 'button' && item.title === unlinkTitle,
      );

      expect(unlinkButton).toBeDefined();
      expect(unlinkButton).toMatchObject({
        icon: UnlinkIcon,
      });
    });

    it('has a remove button for blockCard', () => {
      const { editorView } = editor(
        doc(
          '{<node>}',
          blockCard({
            url: 'http://www.atlassian.com/',
          })(),
        ),
      );

      const toolbar = floatingToolbar({})(
        editorView.state,
        intl,
        providerFactory,
      );
      expect(toolbar).toBeDefined();

      const removeButton = getToolbarItems(toolbar!, editorView).find(
        item => item.type === 'button' && item.title === removeTitle,
      );

      expect(removeButton).toBeDefined();
      expect(removeButton).toMatchObject({
        appearance: 'danger',
        icon: RemoveIcon,
      });
    });

    it('has a remove button for embedCard', () => {
      const { editorView } = editor(
        doc(
          '{<node>}',
          embedCard({
            url: 'http://www.atlassian.com/',
            layout: 'center',
          })(),
        ),
      );

      const toolbar = floatingToolbar({})(
        editorView.state,
        intl,
        providerFactory,
      );
      expect(toolbar).toBeDefined();

      const removeButton = getToolbarItems(toolbar!, editorView).find(
        item => item.type === 'button' && item.title === removeTitle,
      );

      expect(removeButton).toBeDefined();
      expect(removeButton).toMatchObject({
        appearance: 'danger',
        icon: RemoveIcon,
      });
    });

    it('has a visit button', () => {
      const { editorView } = editor(
        doc(
          p(
            '{<node>}',
            inlineCard({
              url: 'http://www.atlassian.com/',
            })(),
          ),
        ),
      );

      const toolbar = floatingToolbar({})(
        editorView.state,
        intl,
        providerFactory,
      );
      expect(toolbar).toBeDefined();

      const visitButton = getToolbarItems(toolbar!, editorView).find(
        item => item.type === 'button' && item.title === visitTitle,
      );

      expect(visitButton).toBeDefined();
      expect(visitButton).toMatchObject({
        icon: OpenIcon,
      });
    });

    it('opens the url in a new window defined on an inline card', () => {
      // @ts-ignore
      global.open = jest.fn();

      const { editorView } = editor(
        doc(
          p(
            '{<node>}',
            inlineCard({
              url: 'http://www.atlassian.com/',
            })(),
          ),
        ),
      );

      const toolbar = floatingToolbar({})(
        editorView.state,
        intl,
        providerFactory,
      );

      const visitButton = getToolbarItems(toolbar!, editorView).find(
        item => item.type === 'button' && item.title === visitTitle,
      ) as FloatingToolbarButton<Command>;

      visitButton.onClick(editorView.state, editorView.dispatch);
      expect(open).toBeCalledWith('http://www.atlassian.com/');
    });

    it('opens the url in a new window via data on an inline card', () => {
      // @ts-ignore
      global.open = jest.fn();

      const { editorView } = editor(
        doc(
          p(
            '{<node>}',
            inlineCard({
              data: {
                url: 'http://www.atlassian.com/',
              },
            })(),
          ),
        ),
      );

      const toolbar = floatingToolbar({})(
        editorView.state,
        intl,
        providerFactory,
      );
      const visitButton = getToolbarItems(toolbar!, editorView).find(
        item => item.type === 'button' && item.title === visitTitle,
      ) as FloatingToolbarButton<Command>;

      visitButton.onClick(editorView.state, editorView.dispatch);
      expect(open).toBeCalledWith('http://www.atlassian.com/');
    });

    it('deletes a block card', () => {
      const { editorView } = editor(
        doc(
          p('ab'),
          '{<node>}',
          blockCard({
            url: 'http://www.atlassian.com/',
          })(),
          p('cd'),
        ),
      );

      const toolbar = floatingToolbar({})(
        editorView.state,
        intl,
        providerFactory,
      );
      const removeButton = getToolbarItems(toolbar!, editorView).find(
        item => item.type === 'button' && item.title === removeTitle,
      ) as FloatingToolbarButton<Command>;

      removeButton.onClick(editorView.state, editorView.dispatch);
      expect(editorView.state.doc).toEqualDocument(doc(p('ab'), p('cd')));
    });

    it('deletes an inline card', () => {
      const { editorView } = editor(
        doc(
          p(
            'ab',
            '{<node>}',
            inlineCard({
              data: {
                title: 'Welcome to Atlassian!',
                url: 'http://www.atlassian.com/',
              },
            })(),
            'cd',
          ),
        ),
      );

      const toolbar = floatingToolbar({})(
        editorView.state,
        intl,
        providerFactory,
      );
      const unlinkButton = getToolbarItems(toolbar!, editorView).find(
        item => item.type === 'button' && item.title === unlinkTitle,
      ) as FloatingToolbarButton<Command>;

      unlinkButton.onClick(editorView.state, editorView.dispatch);
      expect(editorView.state.doc).toEqualDocument(
        doc(p('abWelcome to Atlassian!cd')),
      );
    });

    it('deletes an embed card', () => {
      const { editorView } = editor(
        doc(
          p('ab'),
          '{<node>}',
          embedCard({
            url: 'http://www.atlassian.com/',
            layout: 'center',
          })(),
          p('cd'),
        ),
      );

      const toolbar = floatingToolbar({})(
        editorView.state,
        intl,
        providerFactory,
      );
      const removeButton = getToolbarItems(toolbar!, editorView).find(
        item => item.type === 'button' && item.title === removeTitle,
      ) as FloatingToolbarButton<Command>;

      removeButton.onClick(editorView.state, editorView.dispatch);
      expect(editorView.state.doc).toEqualDocument(doc(p('ab'), p('cd')));
    });

    it('has no toolbar items when url via url attr is invalid', () => {
      const { editorView } = editor(
        doc(
          p(
            '{<node>}',
            inlineCard({
              url: 'javascript:alert(document.domain)',
            })(),
          ),
        ),
      );

      const toolbar = floatingToolbar({})(
        editorView.state,
        intl,
        providerFactory,
      );
      expect(getToolbarItems(toolbar!, editorView).length).toEqual(0);
    });

    it('has no toolbar items when url via data attr is invalid', () => {
      const { editorView } = editor(
        doc(
          p(
            '{<node>}',
            inlineCard({
              url: 'javascript:alert(document.domain)',
            })(),
          ),
        ),
      );

      const toolbar = floatingToolbar({})(
        editorView.state,
        intl,
        providerFactory,
      );
      expect(getToolbarItems(toolbar!, editorView).length).toEqual(0);
    });
  });
});
