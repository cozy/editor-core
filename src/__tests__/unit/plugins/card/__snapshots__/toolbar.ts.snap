// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`card toolbar displays toolbar items in correct order for blockCard 1`] = `
Array [
  Object {
    "render": [Function],
    "type": "custom",
  },
  Object {
    "type": "separator",
  },
  Object {
    "onClick": [Function],
    "selected": false,
    "showTitle": true,
    "title": "Edit link",
    "type": "button",
  },
  Object {
    "type": "separator",
  },
  Object {
    "className": "hyperlink-open-link",
    "icon": [Function],
    "onClick": [Function],
    "title": "Open link in a new tab",
    "type": "button",
  },
  Object {
    "type": "separator",
  },
  Object {
    "appearance": "danger",
    "icon": [Function],
    "onClick": [Function],
    "onMouseEnter": [Function],
    "onMouseLeave": [Function],
    "title": "Remove",
    "type": "button",
  },
]
`;

exports[`card toolbar displays toolbar items in correct order for embedCard 1`] = `
Array [
  Object {
    "render": [Function],
    "type": "custom",
  },
  Object {
    "type": "separator",
  },
  Object {
    "icon": [Function],
    "onClick": [Function],
    "selected": false,
    "title": "Align left",
    "type": "button",
  },
  Object {
    "icon": [Function],
    "onClick": [Function],
    "selected": true,
    "title": "Align center",
    "type": "button",
  },
  Object {
    "icon": [Function],
    "onClick": [Function],
    "selected": false,
    "title": "Align right",
    "type": "button",
  },
  Object {
    "type": "separator",
  },
  Object {
    "icon": [Function],
    "onClick": [Function],
    "selected": false,
    "title": "Wrap left",
    "type": "button",
  },
  Object {
    "icon": [Function],
    "onClick": [Function],
    "selected": false,
    "title": "Wrap right",
    "type": "button",
  },
  Object {
    "type": "separator",
  },
  Object {
    "onClick": [Function],
    "selected": false,
    "showTitle": true,
    "title": "Edit link",
    "type": "button",
  },
  Object {
    "type": "separator",
  },
  Object {
    "className": "hyperlink-open-link",
    "icon": [Function],
    "onClick": [Function],
    "title": "Open link in a new tab",
    "type": "button",
  },
  Object {
    "type": "separator",
  },
  Object {
    "appearance": "danger",
    "icon": [Function],
    "onClick": [Function],
    "onMouseEnter": [Function],
    "onMouseLeave": [Function],
    "title": "Remove",
    "type": "button",
  },
]
`;

exports[`card toolbar displays toolbar items in correct order for inlineCard 1`] = `
Array [
  Object {
    "render": [Function],
    "type": "custom",
  },
  Object {
    "type": "separator",
  },
  Object {
    "onClick": [Function],
    "selected": false,
    "showTitle": true,
    "title": "Edit link",
    "type": "button",
  },
  Object {
    "type": "separator",
  },
  Object {
    "className": "hyperlink-open-link",
    "icon": [Function],
    "onClick": [Function],
    "title": "Open link in a new tab",
    "type": "button",
  },
  Object {
    "type": "separator",
  },
  Object {
    "icon": [Function],
    "onClick": [Function],
    "title": "Unlink",
    "type": "button",
  },
]
`;
